app/services/chart_builder.rb

```diff
@@ -1,15 +1,36 @@
 class ChartBuilder
-  def initialize(benchmark_runs)
+  attr_reader :data, :versions, :benchmark_result_type
+
+  def initialize(benchmark_runs = [], benchmark_result_type)
+    @benchmark_result_type = benchmark_result_type
     @benchmark_runs = benchmark_runs
+    # @versions is a hash that holds all version/environment information for each datapoint, 
+    #   and gets converted to html and stored in @data[:categories]
+    @versions = []
+    # @data[:columns] is JSON that looks like [{ name: "benchmark1", data: [1.1, 1.2] }]
+    # @data[:categories] is JSON that looks like [version_html_string, version_html_string]
     @data = {}
     @data[:columns] = {}
   end
 
+  # `cache_read` looks like 
+  # { datasets: [{ name: "benchmark1", data: [1.1, 1.2] }], versions: [version_hash, version_hash] }
+  def construct_from_cache(cache_read)
+    @versions = cache_read[:versions]
+    @data[:categories] = @versions.map { |version| hash_to_html(version) }.to_json
```

Comment:
Does `hash_to_html` make sense for JSON? Don't be afraid the make `hash_to_html` a view helper method and call it directly in the HTML instead if prescribing all formats to this